proposals/0046-astro-env.md

```diff
@@ -0,0 +1,366 @@
+
+```ts
+type StringField = {
+  type: "string"
+  optional?: boolean
+  default?: string
+}
+
+type NumberField = {
+  type: "number"
+  optional?: boolean
+  default?: number
+}
+
+type NumberField = {
+  type: "boolean"
+  optional?: boolean
+  default?: boolean
+}
+
+type EnvFieldType = StringField | NumberField | BooleanField
+
+type PublicStaticEnvFieldMetadata = {
+  scope: "static"
+  access: "public"
+}
+
+type PrivateStaticEnvFieldMetadata = {
+  scope: "static"
+  access: "private"
+}
+
+type PrivateDynamicEnvFieldMetadata = {
+  scope: "dynamic"
+  access: "private"
+}
+
+type EnvSchema =
+  | Record<`PUBLIC_${string}`, PublicStaticEnvFieldMetadata & EnvFieldType>
+  | Record<string, (PrivateStaticEnvFieldMetadata | PrivateDynamicEnvFieldMetadata) & EnvFieldType>
+
+type AstroUserConfig = {
+  env?: {
+    schema: EnvSchema
+  }
+}
+```
+
+Since there are various combinations possible that can make choosing confusing, we provide a `envField` helper that uses chaining, mirroring other APIs like Astro DB or zod:
+
+```ts
+import { envField } from "astro/config"
+
+// { scope: "static", access: "public", type: "number", default: 4321 }
+envField.static().public().number({ default: 4321 })
+
+// { scope: "dynamic", access: "private", type: "string" }
+envField.dynamic().private().string()
+```
+
+## Integrations
+
+The above way of declaring the schema allows integrations to declare their own constraints from inside `astro:config:setup` without any special handling:
+
+```ts
+import { envField } from "astro/config"
+import type { AstroIntegration } from "astro"
+
+const integration = {
+  name: "supabase-integration",
+  hooks: {
+    "astro:config:setup": (params) => {
+      params.updateConfig({
+        env: {
+          schema: {
+            PUBLIC_SUPABASE_URL: envField.static().public().string()
+          }
+        }
+      })
+    }
+  }
+} satisfies AstroIntegration
+```
+
+## Static variables
+
+Static variables are checked at some point between `astro:config:setup` and `astro:server:setup`. Since we don't/won't support many data structures, custom validators are used for the validation. They will also be used for dynamic variables at runtime.
+
+Some names must be reserved, like `SSR` to avoid conflicts with `import.meta.env.SSR`.
+
+The `astro:env/static` virtual module is generated like so:
+
+```ts
+export const FOO = import.meta.env["FOO"]
+export const PUBLIC_BAR = import.meta.env["PUBLIC_BAR"] ?? true
+```
+
+Types are generated via codegen, likely at `.astro/types.d.ts`:
+
+```ts
+declare module "astro:env/static" {
+  export const FOO: string
+  export const PUBLIC_BAR: boolean
+} 
+```
+
+There's no protection to put in place regarding client-side usage, since `import.meta.env` already handles it.
+
+## Dynamic variables
+
+Dynamic variables are an adapter feature, meaning it's not supported by purely static site (ie. no adapter set).
```

Comment:
That's a bit of an unexpected limitation that would make migrating between `static <> hybrid <> server` more difficult. Any reason _not_ to support it for static builds and use similar behavior to the Node adapter?

Reply:
> use similar behavior to the Node adapter

What's the current behavior? Haven't used it much myself! But I guess runtime features could still be available and just default to node handling, so `process.env`. Is that what you are talking about?

Reply:
I also think we should support this for static builds. It would be kinda like the node adapter and source from `process.env.*` in build time. That means it kinda works like static but we don't prevent still using as dynamic so they can switch the output freely.